core/src/main/java/org/geysermc/geyser/registry/populator/ItemRegistryPopulator.java

```diff
@@ -492,7 +600,28 @@ public static void populate() {
 
                 componentBuilder.putCompound("item_properties", itemProperties.build());
                 builder.putCompound("components", componentBuilder.build());
-                furnaceMinecartData = new ComponentItemData("geysermc:furnace_minecart", builder.build());
+                componentItemData.add(new ComponentItemData("geysermc:furnace_minecart", builder.build()));
+
+                // Register any completely custom items given to us
+                for (NonVanillaCustomItemData customItem : nonVanillaCustomItems) {
+                    int customItemId = nextFreeBedrockId++;
+                    NonVanillaItemRegistration registration = CustomItemRegistryPopulator.registerCustomItem(customItem, customItemId);
+
+                    componentItemData.add(registration.componentItemData());
+                    ItemMapping mapping = registration.mapping();
+                    while (mapping.getJavaId() > mappings.size()) {
+                        // Fill with empty to get to the correct size
+                        mappings.add(ItemMapping.AIR);
+                    }
+                    mappings.add(mapping);
```

Comment:
This doesn't seem right, shouldn't you set the mapping with the Java ID being the index?

Reply:
I'm not sure I understand.

Reply:
`mappings` is a list of items ordered by the item's Java ID. If you register a non-vanilla item with Java Id 2000, then an item with Java Id 1500. Both items would be put at the end of the list and the second item would be in the wrong spot.
```suggestion
                    mappings.set(mapping.getJavaId(), mapping);
```
and change `while (mapping.getJavaId() > mappings.size()) {` to `while (mapping.getJavaId() >= mappings.size()) {`

You may also need to add a check in `registerCustomItem` prevent extensions from registering non vanilla items with duplicate Java IDs.